### PR TITLE
C#: Add ParseFrom/MergeFrom ReadOnlySpan<byte>

### PR DESCRIPTION
**Pull request by Jensaarai**
Address #7885

### CODE REVIEW

File: csharp/src/Google.Protobuf.Test/CodedInputStreamTest.cs [hunk 1]
```diff
@@ -926,6 +926,39 @@ public void TestParseMessagesOver2G()
             }
         }
 
+        [Test]
```
Comment (jtattermusch) [inline csharp/src/Google.Protobuf.Test/CodedInputStreamTest.cs:929]: I'm not sure these test really belong under CodedInputStreamTest.  There's probably a better place where to put them?

Comment (jtattermusch) [reply csharp/src/Google.Protobuf.Test/CodedInputStreamTest.cs:929]: Also, by adding only a simple test like this you will only get a very basic testing. 

If you added a ReadOnlySpan codepath here: https://github.com/protocolbuffers/protobuf/blob/5e84a6169cf0f9716c9285c95c860bcb355dbdc1/csharp/src/Google.Protobuf.Test/MessageParsingHelpers.cs#L53 you'd get a much better coverage and then the Span-based overloads would automatically get a lot of test excercising them.

**Comment by jtattermusch** [comment]
CC @JamesNK  in case he want to review as well.

File: csharp/src/Google.Protobuf/ParseContext.cs [hunk 2]
```diff
@@ -104,6 +104,29 @@ internal static void Initialize(ReadOnlySequence<byte> input, int recursionLimit
             ctx.state.ExtensionRegistry = null;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void Initialize(ReadOnlySpan<byte> input, out ParseContext ctx)
```
[the remark this answers is not in the document]
Comment (Jensaarai) [reply csharp/src/Google.Protobuf/ParseContext.cs]: Done!

File: csharp/src/Google.Protobuf.Test/CodedInputStreamTest.cs (hunk 1 above)
Comment (Jensaarai) [reply csharp/src/Google.Protobuf.Test/CodedInputStreamTest.cs]: Done!

**Review by jtattermusch — APPROVED** [review]
Thanks for addressing my comments.

I actually removed the "ref" from "ref ReadOnlySpan<byte>" for public methods since what I had in mind is to only add "ref" in the internal APIs. It seems that in other public APIs in .NET,  only "ReadOnlySpan<byte>" is usually passed (I guess that's because of readability and safety), we should probably to the same.

Besides that, I made a few readability improvements.

I'm marking the PR as "approved".- please let me know if the changes looks good to you.

**Comment by jtattermusch** [comment]
@JamesNK  can you please take a look as well? Since this is adding a few new public methods, I wanted to get a second opinion.

**Comment by Jensaarai** [comment]
Changes are fine with me :)

**Comment by jtattermusch** [comment]
@JamesNK  please let me know if you're planning to review this or not.

File: csharp/src/Google.Protobuf/MessageExtensions.cs [hunk 3]
```diff
@@ -294,6 +303,16 @@ internal static void MergeFrom(this IMessage message, ReadOnlySequence<byte> dat
             ParsingPrimitivesMessages.CheckReadEndOfStreamTag(ref ctx.state);
         }
 
+        [SecuritySafeCritical]
+        internal static void MergeFrom(this IMessage message, ref ReadOnlySpan<byte> data, bool discardUnknownFields, ExtensionRegistry registry)
```
Comment (JamesNK) [inline csharp/src/Google.Protobuf/MessageExtensions.cs:307]: Does data need to be ref here?

File: csharp/src/Google.Protobuf/ParseContext.cs [hunk 4]
```diff
@@ -65,6 +65,29 @@ internal static void Initialize(ref ReadOnlySpan<byte> buffer, ref ParserInterna
...
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void Initialize(ref ReadOnlySpan<byte> input, out ParseContext ctx)
+        {
+            Initialize(ref input, DefaultRecursionLimit, out ctx);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void Initialize(ref ReadOnlySpan<byte> input, int recursionLimit, out ParseContext ctx)
+        {
```
Comment (JamesNK) [inline csharp/src/Google.Protobuf/ParseContext.cs:76]: Does input need to be ref here?

I think comments explaining how and why these Initialize overloads are different from the existing overload that takes a ReadOnlySpan on line 62.

`internal static void Initialize(ref ReadOnlySpan<byte> buffer, ref ParserInternalState state, out ParseContext ctx)`

**Review by JamesNK — COMMENTED** [review]
Is there a test of what happens if the input ReadOnlySpan doesn't contain enough data? For example, with a ReadOnlySequence we attempt to get more data from the sequence. Does a ReadOnlySpan fail with an expected error in that situation?

**Comment by Jensaarai** [comment]
> Is there a test of what happens if the input ReadOnlySpan doesn't contain enough data? For example, with a ReadOnlySequence we attempt to get more data from the sequence.
> 
> Is there the correct code to handle that situation when there isn't a sequence backing up the span? Does a ReadOnlySpan fail with an expected error in that situation?

I updated the tests around `ReadOnlySequence` to also test the sequence as a `ReadOnlySpan` (after converting to a `byte[]` with `ToArray()`). Do you think this satisfies these cases?

File: csharp/src/Google.Protobuf/MessageExtensions.cs (hunk 3 above)
Comment (Jensaarai) [reply csharp/src/Google.Protobuf/MessageExtensions.cs]: I think you're right, I removed passing `ReadOnlySpan` by ref in these methods and in `ParseContext`...there are more throughout the codebase, e.g. in `ParsingPrimitives` that probably should also change, but I didn't do that here.

I believe `ReadOnlySpan` is not a big enough struct to benefit from any potential performance increase from passing by `ref` or `in` (which is probably why the Microsoft APIs using it don't use `in`)?

https://docs.microsoft.com/en-us/dotnet/csharp/write-safe-efficient-code#apply-the-in-modifier-to-readonly-struct-parameters-larger-than-systemintptrsize

File: csharp/src/Google.Protobuf/ParseContext.cs (hunk 4 above)
Comment (Jensaarai) [reply csharp/src/Google.Protobuf/ParseContext.cs]: As above, I removed passing by `ref` and added some comments with my best understanding of how they were being used!

**Comment by Jensaarai** [comment]
@JamesNK @jtattermusch big enough changes in the last commit probably warrant a re-check?

**Review by jtattermusch — APPROVED** [review]
Still LGTM.

**Comment by jtattermusch** [comment]
@JamesNK  if you're happy with the current version, I will merge.

**Comment by Jensaarai** [comment]
@JamesNK I'm sure you're busy, feel free to tell me to bug off if you queue these up :)

**Review by JamesNK — APPROVED** [review]
LGTM!

**Comment by Jensaarai** [comment]
Ready @jtattermusch ? I'm not sure what to do about the Bagel failure...

**Comment by jtattermusch** [comment]
The bazel failure is unrelated to this PR.

**Pull request merged**